### PR TITLE
Fill subscription id in url param of rest modules automatically

### PR DESCRIPTION
**Pull request by zikalino**
##### SUMMARY
This is small improvement in handling REST API url parameter.
User will be able to use * instead of subscription id, so current subscription id will be filled automatically.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
azure_rm_resource
azure_rm_resource_facts

##### ANSIBLE VERSION
2.6

### CODE REVIEW

File: lib/ansible/modules/cloud/azure/azure_rm_resource.py [hunk 1]
```diff
@@ -27,6 +27,7 @@
   url:
     description:
       - Azure RM Resource URL.
+      - Use * instead of subscription id and it will be replaced with default subscription.
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_resource.py:30]: need be care when using * in string. actually my question is, is filling subscription id in resource url a real problem for user? normally user copy resource url from somewhere else, what's the motivation behind the change.

Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_resource.py:30]: well, resource is known by ansible and comes from environment, so there may be different ways to get it:
- fill it manually
- do the lookup for environment variable
- anything else depending on the environment

But ansible modules just have it, so user won't need to worry about it.


I mean , for other tasks, you don't have to specify subscription id, so if you use URL in _resource module, it may need additional effort from the user, and this change is to avoid it.
Also users may not want to check in their subscription id into their playbooks.

Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_resource.py:30]: I could change "*" to, say, "{subscription}" or "{}", I don't know. but * seems a kind of obvious.

Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_resource.py:30]: whether it's real problem.... yes it may cause some difficulties, there's no single lookup for subscription id, so writing a playbook that works in any environment may be not possible / difficult. we encountered that while working on OpenShift for instance.

**Comment by Fred-sun** [comment]
@zikalino Please help confirm, does this PR ready for review?

**Comment by Fred-sun** [comment]
@zikalino  Please help to confirm the PR ready for review or not?

File: lib/ansible/modules/cloud/azure/azure_rm_resource.py (hunk 1 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_resource.py:30]: second think on this problem. a normal resource url is like:
/subscriptions/<subs_id>/resourceGroups/xxxx/providers/providername/subtype/<resourcename>.

User will meet the problem to get subscription id when manually constructing this long string.  how about make url either be:
1. a url user copied from some one else
2. a dict contains resource group and resource name and provider_name, then module construct url for them.
3. name only. then construct  url from current resource group.

this follows behavior of exisiting azure modules handles referenced resources, like nic in vm module, and also avoid "*"

**Review by yungezz — CHANGES_REQUESTED** [review]
suggest to refine url to be either string or dict.

**Comment by Fred-sun** [comment]
@zikalino Could recheck this PR according by @yungezz's comments? Thanks!

**Comment by Fred-sun** [comment]
@zikalino Could recheck this PR according by @yungezz's comments? Thanks!

**Comment by Fred-sun** [comment]
@zikalino Could you help to recheck this PR according by @yungezz‘s comments? if you have other ideas , please share them with us.  Thanks!!

**Comment by Fred-sun** [comment]
@zikalino Could you help recheck this PR? @yungezze have suggested some changes to refine url? Could you help support it? Thanks!

**Comment by Fred-sun** [comment]
@zikalino

**Comment by Fred-sun** [comment]
@zikalino

**Comment by Fred-sun** [comment]
@zikalino

**Comment by Fred-sun** [comment]
@zikalino need you recheck and update.

**Comment by Fred-sun** [comment]
@zikalino

**Comment by Fred-sun** [comment]
@zikalino Please help update according by yungezz's comment? Thanks!

**Comment by Fred-sun** [comment]
@zikalino Could you help resolve the conflicting files and update according by yungezz's comment? Thanks!

**Comment by Fred-sun** [comment]
@zikalino

**Comment by Fred-sun** [comment]
@zikalino

**Comment by Fred-sun** [comment]
@zikalino Could you help resolve the conflicting files and update according by yungezz's comment? Thanks!

**Comment by Fred-sun** [comment]
@zikalino

**Comment by Fred-sun** [comment]
@zikalino Please help resolve the conflicting files and update according by yungezz's comment! Thanks!

**Comment by Fred-sun** [comment]
@zikalino

**Comment by zikalino** [comment]
closing. decided not to include this feature.
